src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -483,30 +487,56 @@ void ObjectSynchronizer::enter(Handle obj, BasicLock* lock, JavaThread* current)
   current->inc_held_monitor_count();
 
   if (!useHeavyMonitors()) {
-    markWord mark = obj->mark();
-    if (mark.is_neutral()) {
-      // Anticipate successful CAS -- the ST of the displaced mark must
-      // be visible <= the ST performed by the CAS.
-      lock->set_displaced_header(mark);
-      if (mark == obj()->cas_set_mark(markWord::from_pointer(lock), mark)) {
+    if (UseFastLocking) {
+      // fast-locking does not use the 'lock' parameter.
+      LockStack& lock_stack = current->lock_stack();
+      if (lock_stack.can_push()) {
+        markWord header = obj()->mark_acquire();
+        while (true) {
+          if (header.is_neutral()) {
+            assert(!lock_stack.contains(obj()), "thread must not already hold the lock");
+            // Try to swing into 'fast-locked' state without inflating.
+            markWord locked_header = header.set_fast_locked();
+            markWord witness = obj()->cas_set_mark(locked_header, header);
+            if (witness == header) {
+              // Successfully fast-locked, push object to lock-stack and return.
+              lock_stack.push(obj());
+              return;
+            }
+            // Otherwise retry.
+            header = witness;
+          } else {
+            // Fall-through to inflate-enter.
+            break;
+          }
+        }
+      }
+    } else {
```

Comment:
Consider adding a comment after L513 and before L514:
`      // No room on the lock_stack so fall-through to inflate-enter.`